File: Archived2/Java/Gold/1753.java
import java.io.*;
import java.util.*;


public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static ArrayList<ArrayList<Vertex>> adj;
    private static int[] dist;
    private static int startVertex;

    public static void main(String[] args) throws IOException {

        settingInput();
        dijkstra();
        StringBuilder sb = new StringBuilder();
        for (int i=1;i<dist.length;i++) if (dist[i] == Integer.MAX_VALUE) sb.append("INF").append("\n"); else sb.append(dist[i]).append("\n");
        System.out.println(sb);
        br.close();
    }

    private static void settingInput() throws IOException {

        int[] VE = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        startVertex = Integer.parseInt(br.readLine());
        adj = new ArrayList<>();
        dist = new int[VE[0]+1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        for (int i=0;i<VE[0]+1;i++) adj.add(new ArrayList<>());
        for (int i=0;i<VE[1];i++) {
            // u, v, w
            int[] edge = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            adj.get(edge[0]).add(new Vertex( edge[1], edge[2] ));
        }

    }

    private static void dijkstra() {

        PriorityQueue<Vertex> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.weight));
        dist[startVertex] = 0;
        pq.offer(new Vertex(startVertex, 0));

        while(!pq.isEmpty()) {
            Vertex currentVertex = pq.poll();
            if (dist[currentVertex.number] < currentVertex.weight) continue;

            adj.get(currentVertex.number).forEach((nextVertex) -> {
                int nextDist = currentVertex.weight + nextVertex.weight;
                if ( dist[nextVertex.number] > nextDist ) {
                    dist[nextVertex.number] = nextDist;
                    pq.offer(new Vertex(nextVertex.number, nextDist));
                }
            });
        }

    }
}

class Vertex {
    int number;
    int weight;

    Vertex(int number, int weight) {
        this.number = number;
        this.weight = weight;
    }
}

File: Kotlin/Gold/1753.kt
import java.io.*;
import java.util.PriorityQueue
import kotlin.collections.ArrayList

lateinit var adj: ArrayList<ArrayList<Vertex>>
lateinit var dist: IntArray
var startVertex = 0
val br = BufferedReader(InputStreamReader(System.`in`))

fun main() {
    settingInput()
    dijkstra()
    println( buildString {
        for (i in 1 until dist.size) {
            if (dist[i] == Int.MAX_VALUE) append("INF").append("\n")
            else append(dist[i]).append("\n")
        }
    })
}

fun settingInput() {
    val VE = br.readLine().split(" ").map { it.toInt() }
    startVertex = br.readLine().toInt()
    adj = ArrayList()
    repeat(VE[0]+1) { adj.add(ArrayList()) }
    repeat(VE[1]) {
        val edge = br.readLine().split(" ").map { it.toInt() }
        adj.get(edge[0]).add(Vertex(edge[1], edge[2]))
    }
    dist = IntArray(VE[0]+1)
    dist.fill(Int.MAX_VALUE)
}

fun dijkstra() {
    val pq = PriorityQueue { a: Vertex, b: Vertex -> a.weight - b.weight }
    dist[startVertex] = 0
    pq.offer(Vertex(startVertex, 0))
    while (pq.isNotEmpty()) {
        val currentVertex = pq.poll()

        if (dist[currentVertex.number] < currentVertex.weight) continue
        adj.get(currentVertex.number).forEach {
            val nextDist = currentVertex.weight + it.weight
            if (dist[it.number] > nextDist) {
                dist[it.number] = nextDist
                pq.offer(Vertex(it.number, nextDist))
            }
        }
    }
}

data class Vertex(
    val number: Int,
    val weight: Int
)